compiletest: add `max-llvm-major-version` directive

jieyouxu opened the pull request:
To complement existing `min-llvm-version` so contributors don't have to use `ignore-llvm-version: 20 - 99` to emulate `max-llvm-major-version: 19`.

Closes #132305.
cc @workingjubilee who suggested this.

### Implementation steps

- [x] 1. Implement the directive (this PR)
- [x] 2. Open an accompanying dev-guide PR to describe the directive (https://github.com/rust-lang/rustc-dev-guide/pull/2129)

r? bootstrap

jieyouxu committed:
compiletest: add `max-llvm-major-version` directive

There's already `min-llvm-version`, and contributors were using
`ignore-llvm-version: 20 - 99` to emulate `max-llvm-major-version: 19`.

Zalathar commented:
Nit: `ignore-llvm-versions` (sic) in one of the commit messages should be `ignore-llvm-version` (singular).

Zalathar commented:
Not mandatory, but consider also editing the ~5 tests that currently use the N-99 pattern, so that the new directive is actually used.

Zalathar commented on src/tools/compiletest/src/header.rs:
```diff
@@ -1490,29 +1490,38 @@ fn ignore_llvm(config: &Config, line: &str) -> IgnoreDecision {
-            // Ignore if actual version is smaller the minimum required
-            // version
+        if let Some(value) = config.parse_name_value_directive(line, "min-llvm-version") {
+            let value = value.trim();
```
Remark: Separate from this PR, we should just make the parse functions trim automatically. I can't imagine a legitimate use-case for directives actually wanting to see untrimmed whitespace.

jieyouxu replied:
I didn't want to touch that in the current directive handling, tbh. I prefer we just redo this entirely.

jieyouxu commented:
Actually one sec, the logic is wrong.

jieyouxu commented:
> Not mandatory, but consider also editing the ~5 tests that currently use the N-99 pattern, so that the new directive is actually used.

Good point, I'll update existing tests.

jieyouxu commented:
Hang on, what in the world is going on here

https://github.com/rust-lang/rust/blob/a9d17627d241645a54c1134a20f1596127fedb60/src/tools/compiletest/src/header.rs#L1116-L1133

jieyouxu commented:
Hang on this is not ready, I need to unfuck `extract_llvm_version` in a separate PR first.

bors commented:
:umbrella: The latest upstream changes (presumably #132317) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu committed:
tests: use `max-llvm-major-version` instead of `ignore-llvm-version` range like `N - 99`

For tests that use `ignore-llvm-version: N - M`, replace that with
`max-llvm-major-version: N-1`.

jieyouxu commented:
Changes since last force push:

- Rebased on top of https://github.com/rust-lang/rust/pull/132315, this PR needs to wait for that to land first.
- Renamed directive from `max-llvm-version` to `max-llvm-major-version` to more accurately reflect intent (i.e. `//@ max-llvm-major-version: 19` will still run the test if llvm version is `19.99.99`).
- Updated tests that use `ignore-llvm-version: $N - $M` to use `max-llvm-major-version: $(N-1)`.
- Added a FIXME linking to #132348 for `tests/codegen/try_question_mark_nop.rs` because that wants something more like `//@ only-llvm-major-version: 19` instead.

Zalathar commented:
Not sure if this is a good idea or not, but another possibility would be to reframe the directive as something like `llvm-version-less-than`.

In practice, I think a lot of the real-world uses of this directive are going to be along the lines of “this doesn't work on LLVM 21”, in which case `llvm-version-less-than: 21` might make more sense? Hard to say.

jieyouxu commented:
> Not sure if this is a good idea or not, but another possibility would be to reframe the directive as something like `llvm-version-less-than`.
> 
> In practice, I think a lot of the real-world uses of this directive are going to be along the lines of “this doesn't work on LLVM 21”, in which case `llvm-version-less-than: 21` might make more sense? Hard to say.

We could also just use a semver version range: `//@ only-llvm-version-range: >=20.0.0`. Since we're using `semver` in the prerequisite PR anyway this can even do stuff like `//@ only-llvm-version-range: >=19 <21`.

jieyouxu commented:
Hm, I feel like I forgot to set this as ready, lol.

@rustbot ready

jieyouxu commented:
r? bootstrap

jieyouxu commented:
r? bootstrap

jieyouxu commented on src/tools/compiletest/src/directive-list.rs:
```diff
@@ -117,6 +117,7 @@ const KNOWN_DIRECTIVE_NAMES: &[&str] = &[
     "incremental",
     "known-bug",
     "llvm-cov-flags",
+    "max-llvm-major-version",
```
Because usually we're interested in changes between **major** LLVM versions (LLVM is semver). For example, for FileCheck-based tests usually we are interested to limit the test on a maximum major LLVM version e.g. 19 (but we still want the test to run on LLVM 19.1, 19.99.99).

jieyouxu replied:
If you look at the assembly tests, this is exactly the property they want: to run tests up to a maximum llvm major version.

Alternatively, we could also offer `//@ llvm-version-less-than`.

onur-ozkan replied:
It seems quite weird that we have `min-llvm-version` which supports semver and we are adding `max-llvm-major-version` which only looks for the major part. It seems like after this, we will need to add another directive when we need to check minor versions? I got the point but it doesn't seem to justify the situation IMO.

jieyouxu replied:
> It seems like after this, we will need to add another directive when we need to check minor versions? I got the point but it doesn't seem to justify the situation IMO.

Such a need for minor version checking has not yet materialized AFAIK. We can always update the directives to better support our use cases if such a need does occur, but needing to constrain on a major llvm version is a problem at present, especially between llvm major version bumps.

onur-ozkan replied:
So why aren’t we doing this now? Just curious, what’s the reason? The cost for doing it doesn't seem much, or am I missing something? It really seems odd that the LLVM versioning interface for min and and directives is entirely different. Wouldn't it be better when they are unified and documented as {min/max}-llvm-version?

nikic replied:
max-llvm-version would be pretty awkward to use, because you'd always have to write it as `max-llvm-version: 19.99.99` or so.

onur-ozkan replied:
No, you can still just write the major part.

nikic replied:
But wouldn't `max-llvm-version: 19` exclude `19.1.0`, for example, because `19.1.0 > 19`? Or is that not how the version comparison works?

onur-ozkan replied:
Ah, got you now. Yeah, in order to ignore 19 completely, you will need to write 19.99.99, which seems not very good.

jieyouxu replied:
Yeah I should've made that more clear. I did that exactly i.e. max-llvm-version in the initial impl, then it occurred to me it is very awkward when I tried to write the unit tests.

onur-ozkan commented:
@bors r+

bors commented:
:pushpin: Commit f41e1cf962d4a5d5aa11f30d80a2706dc2a3aeab has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan f41e1cf962d4a5d5aa11f30d80a2706dc2a3aeab -->
<!-- homu: {"type":"Approved","sha":"f41e1cf962d4a5d5aa11f30d80a2706dc2a3aeab","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #133006) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:pushpin: Commit 91fa16b211957f6d56bd2ec697c36b2301f2690a has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 91fa16b211957f6d56bd2ec697c36b2301f2690a -->
<!-- homu: {"type":"Approved","sha":"91fa16b211957f6d56bd2ec697c36b2301f2690a","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
